### ISSUE TITLE
[Performance] Remove ThemeRegistry class, use simple arrays all the time instead

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#543]
The ThemeRegistry class was introduced in https://www.drupal.org/node/1011614, due to the size of the theme registry exceeding the MySQL max allowed packet. However, there are a number of downsides that were not considered (or perhaps have changed since the implementation). The ThemeRegistry class extends `BackdropCacheArray` (formerly `DrupalCacheArray`), which is a class that implements ArrayIterator and ArrayAccess to act like a normal array, but it is really an object. This has the following downsides:
- Recreating array access mechanisms in user-space code is slow and expensive. When performing theme suggestions (as Views does, see #448), we have to loop over the entire theme registry, which results in thousands of function calls.
- In Backdrop (and Drupal 8), we're increasing the use of theme suggestions because it reduces the overall number of templates, which makes for better themer experience. It also reduces the size of the overall registry every time we combine two templates into a single one. The original report stated that the cause of the large registry was lightbox2 creating a bunch of dynamic templates, e.g. lightbox2-preset-name1.tpl.php, lightbox2-preset-name2.tpl.php. That may have been common in Drupal 6, but in Drupal 7 and later you'd use theme suggestions anyway, resulting a single registry entry for lightbox2.tpl.php and then calling it with `theme('lightbox2__preset_name')`. The double-underscores are automatically converted into suggestions of the same template without additional entries in the theme registry. This keeps the theme registry smaller and prevents a situation where a single module would have hundreds of theme entries.
- The original premise was to prevent max-packet-size being exceeded, but after the registry has been populated over a long enough period of time, the registry has the capacity to reach this number anyway. On a stock install of Backdrop, the theme registry is only 51KB, so reaching a 1MB registry would likely take hundreds of modules. At that point, you'd need to increase the max packet size in MySQL for other reasons besides the theme registry.
- In the entire discussion of introducing the ThemeRegistry class, there was no mention of the CPU impact of the change, which is drastic.

**Comment by quicksketch** [comment]
To demonstrate the impact of the ThemeRegistry class, I just put a single line at the top of `theme_get_registry()`. If $complete = TRUE, then the entire registry is loaded with a traditional array instead of the partially-constructed ThemeRegistry version.

Here's the impact on the front page:
![theme-registry-homepage](https://cloud.githubusercontent.com/assets/100206/5608366/c654f6e2-9436-11e4-9aee-d5fa85735109.png)

And here's the admin/content page, which has been converted to a view:
![theme-registry-admin-content](https://cloud.githubusercontent.com/assets/100206/5608371/d53b9b5c-9436-11e4-8ca7-bd843a1598f3.png)

As the original issue reports, the theme registry does indeed take up an additional memory (400KB for us). However, the performance cost is more than 10%! Although the theme registry will get larger with more modules, so will the number of calls to theme(). I think that we should nix this memory optimization and take the significant CPU gains instead.

**Comment by quicksketch** [comment]
Hm, I think I'm off here. Reproducing these numbers is difficult. If we look at *only* the function call changes, that's the only place where we should have a performance difference. This does result in a lot fewer functions being called, but even with these differences, the maximum amount improvement we should see should be within those functions not being called.

![theme-registry-diff](https://cloud.githubusercontent.com/assets/100206/5608716/9445f3a8-9441-11e4-8433-26fae6e5e026.png)

So these top 6 functions are the only difference in execution, but the total execution time for all of these functions is a mere 20ms, not 120ms. I think the difference we're seeing here was caused by other activity on my machine at the same time. Namely, PHPStorm reindexing my code after I switched between branches. That slowed the whole result down. I'm going to keep an eye on this in the future, and at least shutdown PHPStorm before I start doing A/B testing.

So actual performance improvement more around 1% when I can get two runs about the same and then only checking out the time difference in the changed functions. I don't think this is worth pursuing unless there is a more definitive improvement. Sorry for the false alarm.
